src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java

```diff
@@ -0,0 +1,366 @@
+				re = pos(ru - 1) + _odims[0];
+			}
+			if (allColumns) {
+				nnz += UtilFunctions.computeNnz(_blocks[bi], rb, re - rb) ;
+			}
+			else {
+				for (int ri = rb; ri < re; ri += _odims[0]) {
+				    nnz += UtilFunctions.computeNnz(_blocks[bi], ri + cl, cu - cl);
+				}
+			}
+			rb = 0;
+		}
+		return nnz;
+	}
+
+	@Override
+	public double[] values(int r) {
+		return valuesAt(index(r));
+	}
+	
+	@Override
+	public double[] valuesAt(int bix) {
+		Warnings.warnFullFP64Conversion(_blocks[bix].length() - 1);
+		return DataConverter.toDouble(_blocks[bix], _blocks[bix].length() - 1);
+	}
+
+	@Override
+	public void incr(int r, int c) {
+	    _blocks[index(r)].set(pos(r, c));
+	}
+
+	@Override
+	public void incr(int r, int c, double delta) {
+		if (delta != 0) {
+			_blocks[index(r)].set(pos(r, c));
+		}
+	}
+
+	@Override
+	public DenseBlock set(double v) {
+		boolean bv = v != 0;
+		for (int i = 0; i < _blocks.length - 1; i++) {
+			_blocks[i].set(0, blockSize() * _odims[0], bv);
+		}
+		_blocks[_blocks.length - 1].set(0, blockSize(_blocks.length - 1) * _odims[0], bv);
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(int rl, int ru, int cl, int cu, double v) {
+	    boolean bv = v != 0;
+		boolean allColumns = cl == 0 && cu == _odims[0];
+		int rb = pos(rl);
+		int re = blockSize() * _odims[0];
+		for (int bi = index(rl); bi <= index(ru - 1); bi++) {
+			if (bi == index(ru - 1)) {
+				re = pos(ru - 1) + _odims[0];
+			}
+			if (allColumns) {
+				_blocks[bi].set(rb, re, bv);
+			}
+			else {
+				for (int ri = rb; ri < re; ri += _odims[0]) {
+				    _blocks[bi].set(ri + cl, ri + cu, bv);
+				}
+			}
+			rb = 0;
+		}
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(int r, int c, double v) {
+		_blocks[index(r)].set(pos(r, c), v != 0);
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(int[] ix, double v) {
+		_blocks[index(ix[0])].set(pos(ix), v != 0);
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(int[] ix, String v) {
+		_blocks[index(ix[0])].set(pos(ix), Boolean.parseBoolean(v));
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(int r, double[] v) {
+		int ri = r * _odims[0];
+		for (int i = ri; i < ri + v.length; i++) {
+			_blocks[index(r)].set(i, v[i - ri] != 0);
+		}
+		return this;
+	}
+
+	@Override
+	public DenseBlock set(DenseBlock db) {
+		double[] reuse = db.valuesAt(0);
+		int srcPos = 0;
+		int srcBi = 0;
+		int destPos = 0;
+		int destBi = 0;
+		int finishedRows = 0;
+		// Loop through both blocks (source and destination) and always take the maximum number of rows that were not
+		// yet read from source or not yet written to destination. This procedure is necessary because although the dimensions
+		// should match there is the option that one DenseBlock was reset() and reused old blocks, therefore it is possible
+		// that one DenseBlock uses a block to represent x rows the other one can actually fit x+y rows in a block.
+		while (true) {
+			int length;
+			if (srcPos != 0) {
```

Comment:
Is this really necessary - except for the boolean DenseBlock all of these dense blocks should be aligned as we systematically compute the blocksize. So maybe its a good idea to use the same blocksize for boolean blocks as well because several binary operations would have the same problem.